presto-geospatial/src/main/java/com/facebook/presto/geospatial/GeoFunctions.java

```diff
@@ -0,0 +1,493 @@
+    public static Slice stStartPoint(@SqlType(NAME) Slice input)
+    {
+        OGCGeometry geometry = getGeometry(deserialize(input), "ST_StartPoint", LINE_STRING);
+        MultiPath lines = (MultiPath) geometry.getEsriGeometry();
+        SpatialReference reference = geometry.getEsriSpatialReference();
+        return serialize(OGCGeometry.createFromEsriGeometry(lines.getPoint(0), reference));
+    }
+
+    @Description("Returns the last point of a LINESTRING geometry as a Point")
+    @ScalarFunction("ST_EndPoint")
+    @SqlType(NAME)
+    public static Slice stEndPoint(@SqlType(NAME) Slice input)
+    {
+        OGCGeometry geometry = getGeometry(deserialize(input), "ST_EndPoint", LINE_STRING);
+        MultiPath lines = (MultiPath) geometry.getEsriGeometry();
+        SpatialReference reference = geometry.getEsriSpatialReference();
+        return serialize(OGCGeometry.createFromEsriGeometry(lines.getPoint(lines.getPointCount() - 1), reference));
+    }
+
+    @Description("Return the X coordinate of the point")
+    @ScalarFunction("ST_X")
+    @SqlType(StandardTypes.DOUBLE)
+    public static double stX(@SqlType(NAME) Slice input)
+    {
+        OGCGeometry geometry = getGeometry(deserialize(input), "ST_X", POINT);
+        return ((OGCPoint) geometry).X();
+    }
+
+    @Description("Return the Y coordinate of the point")
+    @ScalarFunction("ST_Y")
+    @SqlType(StandardTypes.DOUBLE)
+    public static double stY(@SqlType(NAME) Slice input)
+    {
+        OGCGeometry geometry = getGeometry(deserialize(input), "ST_Y", POINT);
+        return ((OGCPoint) geometry).Y();
+    }
+
+    @Description("Returns the closure of the combinatorial boundary of this Geometry")
+    @ScalarFunction("ST_Boundary")
+    @SqlType(NAME)
+    public static Slice stBoundary(@SqlType(NAME) Slice input)
+    {
+        OGCGeometry geometry = deserialize(input);
+        OGCGeometry bound = geometry.boundary();
+        return serialize(bound);
+    }
+
+    @Description("Returns the bounding rectangular polygon of a Geometry")
+    @ScalarFunction("ST_Envelope")
+    @SqlType(NAME)
+    public static Slice stEnvelope(@SqlType(NAME) Slice input)
+    {
+        OGCGeometry geometry = deserialize(input);
+        SpatialReference reference = geometry.getEsriSpatialReference();
+        Envelope envelope = getEnvelope(geometry);
+        return serialize(OGCGeometry.createFromEsriGeometry(envelope, reference));
+    }
+
+    @Description("Returns the Geometry value that represents the point set difference of two geometries")
+    @ScalarFunction("ST_Difference")
+    @SqlType(NAME)
+    public static Slice stDifference(@SqlType(NAME) Slice left, @SqlType(NAME) Slice right)
+    {
+        OGCGeometry leftGeometry = deserialize(left);
+        OGCGeometry rightGeometry = deserialize(right);
+        checkArgument(leftGeometry.getEsriSpatialReference().equals(rightGeometry.getEsriSpatialReference()), "Input Geometries Spatial Reference do not match");
```

Comment:
nit: this check is copy-pasted in multiple places; consider moving it into a helper method?

verifySameSR(leftGeometry, rightGeometry);